diff --git a/src/virgl_context.h b/src/virgl_context.h
index 6380e0f0..25bc1060 100644
--- a/src/virgl_context.h
+++ b/src/virgl_context.h
@@ -131,6 +131,13 @@ struct virgl_context {
 
    /* For DRM native contexts, return the device fd: */
    int (*get_device_fd)(struct virgl_context *ctx);
+
+   /* map a resource into a particular address */
+   void * (*resource_map)(struct virgl_context *ctx,
+                          struct virgl_resource *res,
+                          void *addr,
+                          int32_t prot,
+                          int32_t flags);
 };
 
 struct virgl_context_foreach_args {
diff --git a/src/virglrenderer.c b/src/virglrenderer.c
index bcb47caa..bd6097a2 100644
--- a/src/virglrenderer.c
+++ b/src/virglrenderer.c
@@ -1228,6 +1228,7 @@ int virgl_renderer_resource_map(uint32_t res_handle, void **out_map, uint64_t *o
    int ret = 0;
    void *map = NULL;
    uint64_t map_size = 0;
+   struct virgl_context *ctx;
    struct virgl_resource *res = virgl_resource_lookup(res_handle);
    if (!res || res->mapped)
       return -EINVAL;
@@ -1255,6 +1256,13 @@ int virgl_renderer_resource_map(uint32_t res_handle, void **out_map, uint64_t *o
          ret = vkr_allocator_resource_map(res, &map, &map_size);
          break;
       case VIRGL_RESOURCE_OPAQUE_HANDLE:
+         ctx = virgl_context_lookup(res->opaque_handle_context_id);
+         if (!ctx || !ctx->resource_map)
+            return -EINVAL;
+
+         map = ctx->resource_map(ctx, res, NULL, PROT_WRITE | PROT_READ, MAP_SHARED);
+         map_size = res->map_size;
+         break;
       case VIRGL_RESOURCE_FD_INVALID:
          /* Avoid a default case so that -Wswitch will tell us at compile time
           * if a new virgl resource type is added without being handled here.
@@ -1275,6 +1283,45 @@ int virgl_renderer_resource_map(uint32_t res_handle, void **out_map, uint64_t *o
    return ret;
 }
 
+int virgl_renderer_resource_map_fixed(uint32_t res_handle, void *addr)
+{
+   int ret = 0;
+   void *map = NULL;
+   uint64_t map_size = 0;
+   struct virgl_context *ctx;
+   struct virgl_resource *res = virgl_resource_lookup(res_handle);
+
+   if (!res || res->mapped || res->pipe_resource)
+      return -EINVAL;
+
+   switch (res->fd_type) {
+      case VIRGL_RESOURCE_FD_DMABUF:
+      case VIRGL_RESOURCE_FD_SHM:
+         map = mmap(addr, res->map_size, PROT_WRITE | PROT_READ,
+                    MAP_FIXED | MAP_SHARED, res->fd, 0);
+         break;
+      case VIRGL_RESOURCE_OPAQUE_HANDLE:
+         ctx = virgl_context_lookup(res->opaque_handle_context_id);
+         if (!ctx || !ctx->resource_map)
+            return -EINVAL;
+
+         map = ctx->resource_map(ctx, res, addr, PROT_WRITE | PROT_READ,
+                                 MAP_FIXED | MAP_SHARED);
+         break;
+      case VIRGL_RESOURCE_FD_OPAQUE:
+      case VIRGL_RESOURCE_FD_INVALID:
+         /* Avoid a default case so that -Wswitch will tell us at compile time
+          * if a new virgl resource type is added without being handled here.
+          */
+      break;
+   }
+
+   if (!map || map == MAP_FAILED)
+      return -EINVAL;
+
+   return ret;
+}
+
 int virgl_renderer_resource_unmap(uint32_t res_handle)
 {
    TRACE_FUNC();
@@ -1343,6 +1390,8 @@ virgl_renderer_resource_export_blob(uint32_t res_id, uint32_t *fd_type, int *fd)
       *fd_type = VIRGL_RENDERER_BLOB_FD_TYPE_SHM;
       break;
    case VIRGL_RESOURCE_OPAQUE_HANDLE:
+      *fd_type = VIRGL_RENDERER_BLOB_OPAQUE_HANDLE;
+      break;
    case VIRGL_RESOURCE_FD_INVALID:
       /* Avoid a default case so that -Wswitch will tell us at compile time if a
        * new virgl resource type is added without being handled here.
diff --git a/src/virglrenderer.h b/src/virglrenderer.h
index c7ce4e2f..745f5966 100644
--- a/src/virglrenderer.h
+++ b/src/virglrenderer.h
@@ -421,6 +421,7 @@ VIRGL_EXPORT int
 virgl_renderer_resource_create_blob(const struct virgl_renderer_resource_create_blob_args *args);
 
 VIRGL_EXPORT int virgl_renderer_resource_map(uint32_t res_handle, void **map, uint64_t *out_size);
+VIRGL_EXPORT int virgl_renderer_resource_map_fixed(uint32_t res_handle, void *addr);
 
 VIRGL_EXPORT int virgl_renderer_resource_unmap(uint32_t res_handle);
 
@@ -435,6 +436,7 @@ VIRGL_EXPORT int virgl_renderer_resource_get_map_info(uint32_t res_handle, uint3
 #define VIRGL_RENDERER_BLOB_FD_TYPE_DMABUF        0x0001
 #define VIRGL_RENDERER_BLOB_FD_TYPE_OPAQUE        0x0002
 #define VIRGL_RENDERER_BLOB_FD_TYPE_SHM           0x0003
+#define VIRGL_RENDERER_BLOB_OPAQUE_HANDLE         0x0004
 
 VIRGL_EXPORT int
 virgl_renderer_resource_export_blob(uint32_t res_id, uint32_t *fd_type, int *fd);
